File: test/unit/mesh/python/MeshFunction.py
"""Unit tests for MeshFunctions"""

# Copyright (C) 2011 Garth N. Wells
#
# This file is part of DOLFIN.
#
# DOLFIN is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# DOLFIN is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with DOLFIN. If not, see <http://www.gnu.org/licenses/>.
#
# First added:  2011-03-10
# Last changed: 2011-03-10

import unittest
import numpy.random
from dolfin import *

class NamedMeshFunctions(unittest.TestCase):

    def setUp(self):
        #self.names = ["Cell", "Vertex", "Edge", "Face", "Facet"]
        #self.tps = ['int', 'uint', 'bool', 'double']
        self.names = ["Cell", "Vertex", "Edge", "Face", "Facet"]
        self.tps = ['int', 'uint', 'bool', 'double']
        self.mesh = UnitCubeMesh(3, 3, 3)
        self.funcs = {}
        for tp in self.tps:
            for name in self.names:
                self.funcs[(tp, name)] = eval("%sFunction('%s', self.mesh)"%\
                                              (name, tp))

    def test_size(self):
        for tp in self.tps:
            for name in self.names:
                if name is "Vertex":
                    a = self.funcs[(tp, name)].size()
                    b = self.mesh.num_vertices()
                    self.assertEqual(a, b)
                else:
                    a = self.funcs[(tp, name)].size()
                    b = getattr(self.mesh, "num_%ss"%name.lower())()
                    self.assertEqual(a, b)

    def test_access_type(self):
        type_dict = dict(int=int, uint=int, double=float, bool=bool)
        for tp in self.tps:
            for name in self.names:
                self.assertTrue(isinstance(self.funcs[(tp, name)][0], \
                                           type_dict[tp]))

    def test_numpy_access(self):
        for tp in self.tps:
            for name in self.names:
                values = self.funcs[(tp, name)].array()
                values[:] = numpy.random.rand(len(values))
                self.assertTrue(all(values[i]==self.funcs[(tp, name)][i]
                                    for i in xrange(len(values))))

class MeshFunctions(unittest.TestCase):

    def setUp(self):
        self.mesh = UnitCubeMesh(8, 8, 8)
        self.f = MeshFunction('int', self.mesh, 0)

    def testCreate(self):
        """Create MeshFunctions."""
        v = MeshFunction("uint", self.mesh)

        v = MeshFunction("uint", self.mesh, 0)
        self.assertEqual(v.size(), self.mesh.num_vertices())

        v = MeshFunction("uint", self.mesh, 1)
        self.assertEqual(v.size(), self.mesh.num_edges())

        v = MeshFunction("uint", self.mesh, 2)
        self.assertEqual(v.size(), self.mesh.num_facets())

        v = MeshFunction("uint", self.mesh, 3)
        self.assertEqual(v.size(), self.mesh.num_cells())

    def testCreateAssign(self):
        """Create MeshFunctions with value."""
        i = 10
        v = MeshFunction("uint", self.mesh, 0, i)
        self.assertEqual(v.size(), self.mesh.num_vertices())
        self.assertEqual(v[0], i)

        v = MeshFunction("uint", self.mesh, 1, i)
        self.assertEqual(v.size(), self.mesh.num_edges())
        self.assertEqual(v[0], i)

        v = MeshFunction("uint", self.mesh, 2, i)
        self.assertEqual(v.size(), self.mesh.num_facets())
        self.assertEqual(v[0], i)

        v = MeshFunction("uint", self.mesh, 3, i)
        self.assertEqual(v.size(), self.mesh.num_cells())
        self.assertEqual(v[0], i)

    def testAssign(self):
        f = self.f
        f[3] = 10
        v = Vertex(self.mesh, 3)
        self.assertEqual(f[v], 10)

if __name__ == "__main__":
    unittest.main()

File: test/unit/cpp/mesh/MeshFunction.cpp
// Copyright (C) 2013 Garth N. Wells
//
// This file is part of DOLFIN.
//
// DOLFIN is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// DOLFIN is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with DOLFIN. If not, see <http://www.gnu.org/licenses/>.
//
// First added:  2013-05-30
// Last changed:
//
// Unit tests for MeshFunction

#include <dolfin.h>
#include <gtest/gtest.h>

using namespace dolfin;

TEST(MeshFunctions, test_create_from_domains) {
  // Create mesh
  std::shared_ptr<Mesh> mesh(new UnitSquareMesh(3, 3));
  dolfin_assert(mesh);

  const std::size_t D = mesh->topology().dim();

  // Test setting all values
  for (std::size_t d = 0; d <= D; ++d)
  {
    // Create MeshDomains object
    MeshDomains mesh_domains;
    mesh_domains.init(D);

    mesh->init(d);

    // Build mesh domain
    std::map<std::size_t, std::size_t>& domain = mesh_domains.markers(d);
    for (std::size_t i = 0; i < mesh->num_entities(d); ++i)
      domain.insert(std::make_pair(i, i));

    // Create MeshFunction and test values
    MeshFunction<std::size_t> mf(mesh, d, mesh_domains);
    for (std::size_t i = 0; i < mf.size(); ++i)
      ASSERT_EQ(mf[i], i);
  }

  // Test setting some values only
  for (std::size_t d = 0; d <= D; ++d)
  {
    // Create MeshDomains object
    MeshDomains mesh_domains;
    mesh_domains.init(D);

    mesh->init(d);

    // Build mesh domain
    std::map<std::size_t, std::size_t>& domain = mesh_domains.markers(d);
    const std::size_t num_entities = mesh->num_entities(d);
    for (std::size_t i = num_entities/2; i < num_entities; ++i)
      domain.insert(std::make_pair(i, i));

    // Create MeshFunction and test values
    MeshFunction<std::size_t> mf(mesh, d, mesh_domains);
    for (std::size_t i = 0; i < num_entities/2; ++i)
      ASSERT_EQ(mf[i], std::numeric_limits<std::size_t>::max());
    for (std::size_t i = num_entities/2; i < mf.size(); ++i)
      ASSERT_EQ(mf[i], i);
  }
}

// Test all
int main(int argc, char **argv) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}